Simulator/simulator.py

```diff
@@ -95,324 +85,687 @@ def __init__(self, etype: etype, time: datetime, job_id: int):
+    ) -> None:
+        self.dataset = dataset
+        self.NUM_VMS = NUM_VMS
+        self.VM_CPU = VM_CPU
+
+        self.input_trace = input_trace
+        self.training_trace = training_trace
+        self.job_count = input_trace.job_id.count()
+
+        self.next_jobID: int = 0
+        self.current_reserve_jobs: SortedList = SortedList()
+        self.job_schedule_map: Dict[int, str] = {}
+        self.wait_time_map: Dict[int, int] = {}
+        self.wait_queue: deque = deque()
+
+        self.mean_waiting_time: float
+        self.reserved_cost: float
+        self.on_demand_cost: float
+
+        # Cluster
+        self._reserve_nodes: List[Node] = []
+        vm_name = "m5.16xlarge"
+        for i in range(NUM_VMS):
+            n = Node(vm_name, i, VM_CPU)
+            self._reserve_nodes.append(n)
+
+    """
+    Determine which fixed nodes to schedule jobs on
+    If sufficient fixed nodes are unavailable, return an empty dict (must queue or run on on-demand)
+    Otherwise, return dict indicating how many CPUs to reserve on which nodes 
+    """
+
+    def _first_fit_scheduling(self, cpu_req: int) -> Optional[dict]:
+        counter = 0
+        node_dic = {}
+
+        # jobs need to be scheduled across multiple nodes
+        while cpu_req and counter < len(self._reserve_nodes):
+            node = self._reserve_nodes[counter]
+            cpus_scheduled = min(cpu_req, node.idle_cpu)
+            if cpus_scheduled:
+                node_dic[node._nodeID] = cpus_scheduled
+            cpu_req -= cpus_scheduled
+            counter += 1
+
+        if cpu_req == 0:
+            return node_dic
+        else:
+            return None
+
+    """
+    Determines whether sufficient fixed resources are currently available to run job
+    """
+    def _can_schedule(
+            self,
+            cpu_req: int
+    ) -> bool:
+        node_dic = self._first_fit_scheduling(cpu_req)
+        if node_dic:
+            return True
+        else:
+            return False
+
+    """
+    Reserve specified CPUs on specified nodes for a given job
+    Add given job to set of currently running jobs
+    """
+
+    def _schedule(
+            self,
+            start_time: int,
+            cpu_req: int,
+            runtime_sec: int,
+            job_id: int,
+    ) -> None:
+        node_dic = self._first_fit_scheduling(cpu_req)
+        for node_id, CPU_count in node_dic.items():
+            self._reserve_nodes[node_id].schedule_job(CPU_count, job_id)
+        finish_time = start_time + runtime_sec
+        self.current_reserve_jobs.add((finish_time, start_time, job_id, node_dic.keys()))
+
+    """
+    Returns the next event
+    """
+
+    def _get_next_event(self) -> Optional[Event]:
+        next_arr: pd.Series = None
+        next_dep: pd.Series = None
+        next_arr_time = float('inf')
+        next_dep_time = float('inf')
+
+        # Next Arrival
+        if self.next_jobID < self.job_count:
```

Comment:
yeah this might just be confusion over what self.job_count is - self.job_count counts how many jobs are in the trace. the function increments self.next_jobID after each event, so this check is just making sure that we haven't finished all jobs in the trace already (we'd get an error trying to access the job otherwise)

Reply:
this actually helped me identify the bug I was mentioning with the wait queue earlier - realized I needed to update the self.job_count variable when resubmitting jobs under speculative execution. would've taken me forever to figure out, so thanks!

Reply:
Awesome! glad we caught it.